File: BotsSDK/korebot/src/main/java/kore/botssdk/utils/DownloadUtils.java
package kore.botssdk.utils;

import android.app.DownloadManager;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Environment;
import android.os.Handler;
import android.os.Looper;
import android.webkit.URLUtil;
import android.widget.Toast;

import java.net.HttpURLConnection;
import java.net.URL;
import java.util.concurrent.Executors;

import kore.botssdk.R;

public class DownloadUtils {
    public static void downloadFile(Context context, String downloadUrl, DownloadProgressListener listener) {
        ToastUtils.showToast(context, context.getString(R.string.downloading));
        Handler handler = new Handler(Looper.getMainLooper());
        Executors.newSingleThreadExecutor().execute(() -> {
            try {
                URL url = new URL(downloadUrl);
                HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                connection.setRequestMethod("HEAD");
                connection.connect();

                String mimeType = connection.getContentType();
                String contentDisposition = connection.getHeaderField("Content-Disposition");
                connection.disconnect();

                handler.post(() -> download(context, downloadUrl, contentDisposition, mimeType, listener));
            } catch (Exception e) {
                e.printStackTrace();
                handler.post(() -> {
                    ToastUtils.showToast(context, context.getString(R.string.downloading_failed));
                    if (listener != null) listener.onProgress(-1);
                });
            }
        });
    }

    private static void download(Context context, String url, String contentDisposition, String mimeType, DownloadProgressListener listener) {
        DownloadManager.Request request = new DownloadManager.Request(Uri.parse(url));

        String fileName = URLUtil.guessFileName(url, contentDisposition, mimeType);
        request.setTitle(fileName);
        request.setDescription(context.getString(R.string.downloading_file));

        request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);
        request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, "KoreMedia/" + fileName);

        request.setAllowedNetworkTypes(DownloadManager.Request.NETWORK_WIFI | DownloadManager.Request.NETWORK_MOBILE);
        request.allowScanningByMediaScanner();

        DownloadManager downloadManager = (DownloadManager) context.getSystemService(Context.DOWNLOAD_SERVICE);
        long downloadId = downloadManager.enqueue(request);
        ToastUtils.showToast(context, context.getString(R.string.downloading_file_name, fileName), Toast.LENGTH_LONG);
        if (listener != null) {
            Handler handler = new Handler();
            Runnable runnable = new Runnable() {
                @Override
                public void run() {
                    DownloadManager.Query query = new DownloadManager.Query();
                    query.setFilterById(downloadId);
                    Cursor cursor = downloadManager.query(query);

                    if (cursor != null && cursor.moveToFirst()) {
                        int bytesDownloaded = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR));
                        int bytesTotal = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_TOTAL_SIZE_BYTES));

                        if (bytesTotal > 0) {
                            int progress = (int) ((bytesDownloaded * 100L) / bytesTotal);
                            listener.onProgress(progress);
                        }

                        int status = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_STATUS));
                        if (status == DownloadManager.STATUS_SUCCESSFUL || status == DownloadManager.STATUS_FAILED) {
                            cursor.close();
                            return;
                        }
                        cursor.close();
                    }
                    handler.postDelayed(this, 500);
                }
            };

            handler.post(runnable);
        }
    }

    public interface DownloadProgressListener {
        void onProgress(int progress);
    }
}

File: BotSdk/ui/src/main/java/com/kore/ui/utils/DownloadUtils.kt
package com.kore.ui.utils

import android.app.DownloadManager
import android.content.Context
import android.net.Uri
import android.os.Environment
import android.os.Handler
import android.os.Looper
import android.webkit.URLUtil
import android.widget.Toast
import com.kore.ui.R
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.Executors
import androidx.core.net.toUri

object DownloadUtils {
    fun downloadFile(context: Context, downloadUrl: String, listener: DownloadProgressListener?) {
        Toast.makeText(context, context.getString(R.string.downloading), Toast.LENGTH_LONG).show()
        val handler = Handler(Looper.getMainLooper())
        Executors.newSingleThreadExecutor().execute {
            try {
                val url = URL(downloadUrl)
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "HEAD"
                connection.connect()

                val mimeType = connection.contentType
                val contentDisposition = connection.getHeaderField("Content-Disposition")
                connection.disconnect()

                handler.post { download(context, downloadUrl, contentDisposition, mimeType, listener) }
            } catch (e: Exception) {
                e.printStackTrace()
                handler.post {
                    Toast.makeText(context, context.getString(R.string.downloading_failed), Toast.LENGTH_LONG).show()
                    listener?.onProgress(-1)
                }
            }
        }
    }

    private fun download(context: Context, url: String, contentDisposition: String?, mimeType: String?, listener: DownloadProgressListener?) {
        val request = DownloadManager.Request(url.toUri())

        val fileName = URLUtil.guessFileName(url, contentDisposition, mimeType)
        request.setTitle(fileName)
        request.setDescription(context.getString(R.string.downloading_file))

        request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
        request.setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, "KoreMedia/$fileName")

        request.setAllowedNetworkTypes(DownloadManager.Request.NETWORK_WIFI or DownloadManager.Request.NETWORK_MOBILE)
        request.allowScanningByMediaScanner()

        val downloadManager = context.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
        val downloadId = downloadManager.enqueue(request)
        Toast.makeText(context, context.getString(R.string.downloading_file_name, fileName), Toast.LENGTH_LONG).show()
        if (listener != null) {
            val handler = Handler()
            val runnable: Runnable = object : Runnable {
                override fun run() {
                    val query = DownloadManager.Query()
                    query.setFilterById(downloadId)
                    val cursor = downloadManager.query(query)

                    if (cursor != null && cursor.moveToFirst()) {
                        val bytesDownloaded = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR))
                        val bytesTotal = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_TOTAL_SIZE_BYTES))

                        if (bytesTotal > 0) {
                            val progress = ((bytesDownloaded * 100L) / bytesTotal).toInt()
                            listener.onProgress(progress)
                        }

                        val status = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_STATUS))
                        if (status == DownloadManager.STATUS_SUCCESSFUL || status == DownloadManager.STATUS_FAILED) {
                            cursor.close()
                            return
                        }
                        cursor.close()
                    }
                    handler.postDelayed(this, 500)
                }
            }
            handler.post(runnable)
        }
    }

    interface DownloadProgressListener {
        fun onProgress(progress: Int)
    }
}
